File: demo/agent.ts
/**
 * Agento Demo Agent — single agent testing all DeFi tools
 *
 * This agent connects to a running Agento REST server, auto-discovers
 * all available tools via SKILLS.md / GET /tools, and runs through
 * swap, stake, limit order, and lending operations autonomously.
 *
 * Usage:
 *   1. Start the server:  agento serve rest
 *   2. Create + fund a wallet via the CLI
 *   3. Set env vars in .env (WALLET_ID, WALLET_PASSWORD, OPENROUTER_API_KEY)
 *   4. Run:  npm run demo
 */

import { tool } from "@langchain/core/tools";
import { createReactAgent } from "@langchain/langgraph/prebuilt";
import { ChatOpenAI } from "@langchain/openai";
import { HumanMessage } from "@langchain/core/messages";
import { z } from "zod";
import "dotenv/config";

const AGENTO_URL = process.env.AGENTO_URL || "http://localhost:3000";
const WALLET_ID = process.env.WALLET_ID || "";
const WALLET_PASSWORD = process.env.WALLET_PASSWORD || "";

if (!WALLET_ID || !WALLET_PASSWORD) {
  console.error("❌ Set WALLET_ID and WALLET_PASSWORD in .env");
  process.exit(1);
}
if (!process.env.OPENROUTER_API_KEY) {
  console.error("❌ Set OPENROUTER_API_KEY in .env");
  process.exit(1);
}

// ── Health check ──

const health = await fetch(`${AGENTO_URL}/health`).catch(() => null);
if (!health?.ok) {
  console.error(`❌ Cannot reach Agento at ${AGENTO_URL}. Run: agento serve rest`);
  process.exit(1);
}

// ── Discover tools ──

const { tools: toolList } = await (await fetch(`${AGENTO_URL}/tools`)).json() as {
  tools: { name: string; description: string }[];
};

const skillDocs = toolList.map((t) => `• ${t.name}: ${t.description}`).join("\n");
console.log(`🔧 Discovered ${toolList.length} tools from Agento\n`);

// ── Generic Agento caller ──

const agento = tool(
  async ({ toolName, input }) => {
    try {
      const res = await fetch(`${AGENTO_URL}/tools/${toolName}`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "X-Wallet-Id": WALLET_ID,
          "X-Wallet-Password": WALLET_PASSWORD,
        },
        body: JSON.stringify(input ?? {}),
      });
      return JSON.stringify(await res.json());
    } catch (err) {
      return JSON.stringify({ error: String(err) });
    }
  },
  {
    name: "agento",
    description: `Call any Agento wallet tool by name. Available tools:\n${skillDocs}`,
    schema: z.object({
      toolName: z.string().describe("The exact tool name"),
      input: z.record(z.unknown()).optional().describe("Tool input parameters"),
    }),
  },
);

// ── LLM via OpenRouter ──

const llm = new ChatOpenAI({
  modelName: process.env.OPENROUTER_MODEL || "openai/gpt-4o-mini",
  temperature: 0,
  apiKey: process.env.OPENROUTER_API_KEY,
  configuration: { baseURL: "https://openrouter.ai/api/v1" },
});

const agent = createReactAgent({ llm, tools: [agento] });

// ── Test tasks ──

const tasks = [
  {
    name: "💰 Wallet & Prices",
    prompt: "Check my SOL balance, list all token balances, and fetch the current price of SOL and USDC.",
  },
  {
    name: "🔁 Swap SOL → USDC",
    prompt: "Swap 0.1 SOL for USDC. Then check my updated SOL and token balances.",
  },
  {
    name: "🥩 Stake SOL → jupSOL",
    prompt: "Stake 0.1 SOL for jupSOL. Then check my token balances to confirm.",
  },
  {
    name: "📋 Limit Order",
    prompt: `Place a limit order selling 50000000 lamports of SOL (inputMint: So11111111111111111111111111111111111111112) for at least 1000000 USDC units (outputMint: EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v). Use makingAmount and takingAmount as strings. Then list all open orders.`,
  },
  {
    name: "❌ Cancel Orders",
    prompt: "List all my open limit orders. If any exist, cancel all of them.",
  },
  {
    name: "🏦 Lend USDC",
    prompt: "Lend 0.5 USDC (mintAddress: EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v) to Lulo. Then check my token balances.",
  },
  {
    name: "🏧 Withdraw USDC",
    prompt: "Withdraw 0.5 USDC (mintAddress: EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v) from Lulo. Then check my token balances.",
  },
];

// ── Run ──

console.log("🤖 Agento Demo Agent");
console.log(`📡 ${AGENTO_URL}`);
console.log(`💳 Wallet: ${WALLET_ID}`);
console.log(`👀 Dashboard: ${AGENTO_URL}/dashboard`);
console.log(`${"═".repeat(60)}\n`);

const results: { name: string; ok: boolean }[] = [];

for (const task of tasks) {
  console.log(`🧪 ${task.name}`);
  console.log("─".repeat(60));

  try {
    const { messages } = await agent.invoke({
      messages: [new HumanMessage(task.prompt)],
    });
    const reply = messages.at(-1)?.content;
    console.log(`\n🤖 ${typeof reply === "string" ? reply : JSON.stringify(reply)}\n`);
    results.push({ name: task.name, ok: true });
  } catch (err) {
    console.error(`\n❌ ${err instanceof Error ? err.message : err}\n`);
    results.push({ name: task.name, ok: false });
  }
}

console.log("═".repeat(60));
console.log("📊 RESULTS");
console.log("═".repeat(60));
results.forEach((r) => console.log(`${r.ok ? "✅" : "❌"}  ${r.name}`));
console.log(`\n👀 Full activity log: ${AGENTO_URL}/dashboard\n`);

File: examples/python-agent/agent.py
"""
Agento Python Agent — standalone demo

Connects to a hosted Agento REST server, auto-discovers all available tools,
and lets an LLM (via OpenRouter) manage Solana wallets through function calling.

Zero dependency on the Agento Node.js codebase — pure HTTP.

Usage:
    python agent.py                              # default demo
    python agent.py "swap 0.1 SOL for USDC"     # custom task
    python agent.py -i                           # interactive chat
"""

import os
import sys
import json
import httpx
from openai import OpenAI
from dotenv import load_dotenv

load_dotenv()

# ── Config ──

AGENTO_URL = os.environ.get("AGENTO_URL", "https://agento-8m72.onrender.com")
WALLET_ID = os.environ.get("WALLET_ID", "")
WALLET_PASSWORD = os.environ.get("WALLET_PASSWORD", "agento")
MODEL = os.environ.get("MODEL", "openai/gpt-4o-mini")

client = OpenAI(
    api_key=os.environ["OPENROUTER_API_KEY"],
    base_url="https://openrouter.ai/api/v1",
)

http = httpx.Client(timeout=60)

# ── Auto-discover tools from the live API ──

def discover_tools() -> list[dict]:
    """Fetch tool list from Agento and convert to OpenAI function-calling format."""
    resp = http.get(f"{AGENTO_URL}/tools")
    resp.raise_for_status()
    tools = resp.json()["tools"]

    openai_tools = []
    for t in tools:
        openai_tools.append({
            "type": "function",
            "function": {
                "name": t["name"],
                "description": t["description"],
                "parameters": {"type": "object", "properties": {}},
            },
        })
    return openai_tools


# ── REST client ──

wallet_id: str = WALLET_ID


def call_tool(name: str, args: dict) -> dict:
    global wallet_id

    headers: dict[str, str] = {
        "Content-Type": "application/json",
        "X-Wallet-Password": WALLET_PASSWORD,
    }
    if wallet_id:
        headers["X-Wallet-Id"] = wallet_id

    resp = http.post(f"{AGENTO_URL}/tools/{name}", json=args, headers=headers)
    data = resp.json()

    # Auto-capture wallet ID on creation
    if name == "create_wallet" and data.get("success"):
        result = data.get("result", {})
        new_id = result.get("wallet_id") or result.get("id", "")
        if new_id:
            wallet_id = new_id
            print(f"  📝 Active wallet → {wallet_id[:8]}…")

    return data


# ── Agent loop ──

SYSTEM = """You are an AI agent managing Solana wallets through the Agento API.
You operate on Solana devnet. Common token mints:
  SOL:  So11111111111111111111111111111111111111112
  USDC: EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v

Rules:
- If no wallet exists, create one first and airdrop SOL before doing anything.
- Be concise. Report results with addresses and amounts.
- If a tool fails, explain why and suggest a fix.
- The tools accept JSON parameters — pass them as the function arguments."""


def agent_loop(prompt: str, messages: list, tools: list, max_turns: int = 12) -> str | None:
    messages.append({"role": "user", "content": prompt})

    for _ in range(max_turns):
        resp = client.chat.completions.create(
            model=MODEL,
            messages=messages,
            tools=tools,
            tool_choice="auto",
        )

        msg = resp.choices[0].message
        messages.append(msg)

        if not msg.tool_calls:
            print(f"\n🤖 {msg.content}")
            return msg.content

        for tc in msg.tool_calls:
            fn = tc.function.name
            fn_args = json.loads(tc.function.arguments) if tc.function.arguments else {}
            print(f"  🔧 {fn}({json.dumps(fn_args)})")

            result = call_tool(fn, fn_args)
            icon = "✅" if result.get("success") else ("🛡️" if result.get("blocked") else "❌")
            print(f"  {icon} {json.dumps(result)[:200]}")

            messages.append({
                "role": "tool",
                "tool_call_id": tc.id,
                "content": json.dumps(result),
            })

    print("⚠️  Max turns reached.")
    return None


def interactive(tools: list):
    messages: list = [{"role": "system", "content": SYSTEM}]
    print("Type your instructions (ctrl+c to quit):\n")
    while True:
        try:
            prompt = input("You: ").strip()
        except (KeyboardInterrupt, EOFError):
            print("\n👋 Bye.")
            break
        if not prompt:
            continue
        agent_loop(prompt, messages, tools)
        print()


# ── Entry ──

def main():
    print(f"🌐 Agento: {AGENTO_URL}")
    print(f"🧠 Model:  {MODEL}")
    if wallet_id:
        print(f"💳 Wallet: {wallet_id[:8]}…")
    print()

    # Health check
    try:
        resp = http.get(f"{AGENTO_URL}/health")
        resp.raise_for_status()
    except (httpx.ConnectError, httpx.HTTPStatusError):
        print(f"❌ Cannot reach Agento at {AGENTO_URL}")
        sys.exit(1)
    print("✅ Agento is reachable.")

    # Auto-discover tools
    tools = discover_tools()
    print(f"🔧 Discovered {len(tools)} tools.\n")

    if "-i" in sys.argv or "--interactive" in sys.argv:
        interactive(tools)
    else:
        prompt = (
            " ".join(a for a in sys.argv[1:] if not a.startswith("-"))
            or "Create a new wallet, airdrop 2 SOL, then check the balance."
        )
        print(f"🎯 Task: {prompt}\n")
        messages: list = [{"role": "system", "content": SYSTEM}]
        agent_loop(prompt, messages, tools)


if __name__ == "__main__":
    main()
